Automattic/wp-calypso
# Checkout: Prevent plan upgrade variants which are term downgrades

#### Changes proposed in this Pull Request

This PR prevents the checkout plan variant picker from showing plan upgrades which are billing term downgrades because that sort of upgrade is not supported by our payments system. That is, it prevents upgrading from one plan to another plan that has a shorter renewal period. For example, while switching from a Yearly Personal plan to a Yearly Business plan is allowed, switching from a Yearly Personal plan to a Monthly Business plan is _not_ allowed.

To prevent some cases where you could end up switching variants and then not be able to return to your previously selected variant, this also disables the variant picker when the cart item is a variant of the active plan.

Finally, to prevent confusion about what the "Save X%" numbers mean in cases where the monthly price is not shown, this also modifies the discount percentage shown on each product variant, which is currently based on the monthly price for that variant, so that it now is based on the smallest displayed variant's price.

Fixes https://github.com/Automattic/wp-calypso/issues/52473

#### Screenshots

On a site that has an annual premium plan upgrading to a business plan, this is editing the order in checkout before this PR:

<img width="570" alt="Screen Shot 2021-07-08 at 7 52 46 PM" src="https://user-images.githubusercontent.com/2036909/125003848-23447480-e026-11eb-990d-5b50c6ee7a57.png">

And this is editing the same order in checkout after this PR:

<img width="571" alt="Screen Shot 2021-07-08 at 7 52 56 PM" src="https://user-images.githubusercontent.com/2036909/125003854-263f6500-e026-11eb-8a83-bc59e414bc70.png">

Here's editing the checkout order for a site with an (expiring) annual premium plan downgrading to monthly before this PR:

<img width="569" alt="Screen Shot 2021-07-08 at 8 03 48 PM" src="https://user-images.githubusercontent.com/2036909/125004555-c8ac1800-e027-11eb-9a55-98389019195f.png">

And here's editing that same order after this PR:

<img width="574" alt="Screen Shot 2021-07-08 at 8 03 56 PM" src="https://user-images.githubusercontent.com/2036909/125004566-cfd32600-e027-11eb-896e-5aad2db14705.png">


#### Testing instructions

Verify unchanged behavior for free plans:

- Using a site which has a free plan, add a paid plan (of any term length) to your cart and visit checkout.
- Click the "Edit" button at the top-right of the first step.
- Verify that you see a variation picker for the plan and that it includes the monthly, yearly, and two-year plan variants.

Verify that monthly plans can upgrade to any term:

- Next, using a site which has a paid _monthly_ plan, add a higher level _monthly_ plan to your cart and visit checkout (eg: if you have a Personal plan, add a Premium or Business plan).
- Click the "Edit" button at the top-right of the first step.
- Verify that you see a variation picker for the plan and that it includes the monthly, yearly, and two-year plan variants.

Verify that yearly plans cannot upgrade to a monthly term:

- Next, using a site which has a paid _yearly_ plan, add a higher level _yearly_ plan to your cart and visit checkout.
- Click the "Edit" button at the top-right of the first step.
- Verify that you see a variation picker for the plan and that it includes only the yearly and two-year plan variants.
